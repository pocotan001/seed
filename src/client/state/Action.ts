// Flux standard action
// https://github.com/redux-utilities/flux-standard-action
export interface Action<Type, Payload = null, Meta = undefined> {
  type: Type;
  payload: Payload;
  meta?: Meta;
}

export interface ErrorAction<Type, Payload = Error, Meta = undefined> {
  type: Type;
  payload: Payload;
  error: true;
  meta?: Meta;
}
