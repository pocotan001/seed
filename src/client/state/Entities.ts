export interface NormalizedEntities<Entity> {
  [id: number]: Entity;
}
