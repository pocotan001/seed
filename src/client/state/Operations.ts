import { Dispatch } from "redux";
import { Persistor } from "redux-persist";
import { Repositories } from "../infra/repositories";
import createAuthOperations, { AuthOperations } from "./auth/AuthOperations";
import createLoadingOperations, {
  LoadingOperations
} from "./loading/LoadingOperations";
import createRootOperations, { RootOperations } from "./root/RootOperations";

export interface Operations {
  root: RootOperations;
  auth: AuthOperations;
  loading: LoadingOperations;
}

const createOperations = (
  dispatch: Dispatch,
  persistor: Persistor,
  repos: Repositories
): Operations => ({
  root: createRootOperations(dispatch),
  auth: createAuthOperations(dispatch, persistor, repos.auth),
  loading: createLoadingOperations(dispatch)
});

export default createOperations;
