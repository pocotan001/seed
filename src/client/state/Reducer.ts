import { connectRouter } from "connected-react-router";
import { History } from "history";
import { combineReducers, Reducer as ReduxReducer } from "redux";
import { reducer as formReducer } from "redux-form";
import { persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import storageSession from "redux-persist/lib/storage/session";
import authReducer from "./auth/AuthReducer";
import loadingReducer from "./loading/LoadingReducer";
import { State } from "./State";

export type Reducer = ReduxReducer<State>;

const createReducer = (history: History): Reducer =>
  combineReducers({
    router: connectRouter(history),
    form: persistReducer(
      {
        key: "form",
        storage: storageSession,
        whitelist: []
      },
      formReducer
    ),
    auth: persistReducer(
      {
        storage,
        key: "auth"
      },
      authReducer
    ),
    loading: loadingReducer
  });

export default createReducer;
