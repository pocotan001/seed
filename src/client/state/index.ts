export { RootState as State } from "./root/RootState";
export { default } from "./root/RootReducer";
