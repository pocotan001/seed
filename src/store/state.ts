import { createLocation, Location } from "history";
import { extendObservable, observable } from "mobx";
import config from "~/config";
import { INormalizedEntities } from "~/domain/Normalized";

export enum SessionKey {
  EXAMPLE_FORM = "EXAMPLE_FORM"
}

export const defaultState: State = Object.freeze({
  app: {
    locationOrigin: "",
    hasError: false
  },
  history: {
    location: createLocation(""),
    visited: {}
  },
  head: {
    title: config.siteName,
    meta: [],
    link: []
  },
  loading: {
    percent: 0,
    hidden: true
  },
  session: {},
  entities: {
    user: null,
    cats: {}
  },
  results: {
    cats: {}
  },
  cats: {
    totalCount: 0
  }
});

export class State {
  app!: {
    locationOrigin: string;
    hasError: boolean;
  };

  history!: {
    location: Location;
    visited: { [key: string]: boolean };
  };

  head!: {
    title: string;
    meta: Array<React.MetaHTMLAttributes<HTMLMetaElement>>;
    link: Array<React.LinkHTMLAttributes<HTMLLinkElement>>;
  };

  loading!: {
    percent: number;
    hidden: boolean;
  };

  session!: { [K in SessionKey]?: any };

  entities!: INormalizedEntities;

  results!: Record<
    "cats",
    {
      [serializedParams: string]: string[];
    }
  >;

  cats!: {
    totalCount: number;
  };

  constructor(initialState?: Partial<State>) {
    extendObservable(
      this,
      { ...defaultState, ...initialState },
      {},
      {
        defaultDecorator: observable.shallow
      }
    );
  }
}

const createState = (initialState?: Partial<State>) => new State(initialState);

export default createState;
