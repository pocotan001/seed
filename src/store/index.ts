import AppStore from "./AppStore";
import AuthenticationStore from "./AuthenticationStore";
import CatStore from "./CatStore";
import HeadStore from "./HeadStore";
import HistoryStore from "./HistoryStore";
import LoadingStore from "./LoadingStore";
import SessionStore from "./SessionStore";
import { State } from "./state";
import { IStoreContext } from "./Store";

export class RootStore {
  state: State;
  app: AppStore;
  auth: AuthenticationStore;
  head: HeadStore;
  history: HistoryStore;
  loading: LoadingStore;
  session: SessionStore;
  cat: CatStore;

  constructor(state: State, ctx: IStoreContext) {
    this.state = state;
    this.app = new AppStore(state, ctx);
    this.auth = new AuthenticationStore(state, ctx);
    this.head = new HeadStore(state, ctx);
    this.history = new HistoryStore(state, ctx);
    this.loading = new LoadingStore(state, ctx);
    this.session = new SessionStore(state, ctx);
    this.cat = new CatStore(state, ctx);
  }
}

const createStore = (state: State, ctx: IStoreContext) =>
  new RootStore(state, ctx);

export default createStore;
