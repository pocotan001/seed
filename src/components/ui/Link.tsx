import { inject } from "mobx-react";
import * as React from "react";
import styled from "styled-components";
import { Color } from "~/components/styles/theme";
import { RootStore } from "~/store";

type MouseEvent = React.MouseEvent<HTMLAnchorElement>;

export interface LinkProps
  extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  href: string;
  children: React.ReactNode;
  asText?: boolean;
  store?: RootStore;
}

const isAnchorLink = (href: string): boolean => href.indexOf("#") === 0;
const isLeftClick = (e: MouseEvent): boolean => e.button === 0;
const isModifiedEvent = (e: MouseEvent): boolean =>
  Boolean(e.metaKey || e.altKey || e.ctrlKey || e.shiftKey);

@inject("store")
export class Link extends React.PureComponent<LinkProps> {
  store = this.props.store!;

  handleClick = (e: MouseEvent) => {
    const { href, target, onClick } = this.props;

    if (onClick) {
      onClick(e);
    }

    if (
      target ||
      e.defaultPrevented ||
      isAnchorLink(href) ||
      !isLeftClick(e) ||
      isModifiedEvent(e)
    ) {
      return;
    }

    e.preventDefault();

    if (href === this.store.state.history.location.pathname) {
      return;
    }

    this.store.history.push(href);
  };

  render() {
    const { href, children, asText, onClick, store, ...rest } = this.props;

    if (rest.target === "_blank") {
      rest.rel = rest.rel || "noopener noreferrer";
    }

    return (
      <a href={href} onClick={this.handleClick} {...rest}>
        {children}
      </a>
    );
  }
}

export default styled(Link)`
  display: inline-block;
  max-width: 100%;
  text-decoration: none;
  color: currentColor;

  ${({ asText }) =>
    asText &&
    `
    color: ${Color.Pink500};

    &:hover {
      text-decoration: underline;
    }
  `};
`;
