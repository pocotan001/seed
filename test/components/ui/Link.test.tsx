import * as React from "react";
import * as renderer from "react-test-renderer";
import Link from "~/components/ui/Link";
import { RootStore } from "~/store";
import { defaultState } from "~/store/state";

let store: RootStore;

describe("<Link>", () => {
  beforeEach(() => {
    store = {
      state: { ...defaultState },
      history: {
        push: jest.fn()
      }
    } as any;
  });

  it("should render correctly", () => {
    const tree = renderer
      .create(
        <Link href="/a" store={store}>
          alo
        </Link>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should call `history.push` if clicked", () => {
    const tree = renderer
      .create(
        <Link href="/a" store={store}>
          alo
        </Link>
      )
      .toJSON();

    const event: any = {
      button: 0,
      defaultPrevented: false,
      preventDefault: () => undefined
    };

    tree!.props.onClick(event);

    expect(store.history.push).toHaveBeenCalledTimes(1);
    expect(store.history.push).toHaveBeenCalledWith("/a");
    expect(tree).toMatchSnapshot();
  });
});
