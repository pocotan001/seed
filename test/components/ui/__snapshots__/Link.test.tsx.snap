// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link> renders correctly 1`] = `
<a
  className="sc-bdVaJa kSKvLH"
  href="/a"
  onClick={[Function]}
>
  alo
</a>
`;

exports[`<Link> should call \`history.push\` if clicked 1`] = `
<a
  className="sc-bdVaJa kSKvLH"
  href="/a"
  onClick={[Function]}
>
  alo
</a>
`;
